Data/samples/DBLogger/src/DBLogger.cpp

```diff
@@ -0,0 +1,337 @@
+	{
+		Application::defineOptions(options);
+
+		options.addOption(
+			Option("help", "h", "display help information on command line arguments")
+				.required(false)
+				.repeatable(false)
+				.callback(OptionCallback<DBLogger>(this, &DBLogger::handleHelp)));
+
+		options.addOption(
+			Option("?", "?", "display help information on command line arguments")
+				.required(false)
+				.repeatable(false)
+				.callback(OptionCallback<DBLogger>(this, &DBLogger::handleHelp)));
+
+		options.addOption(
+			Option("dir", "d", "directory path to scan for SQL log files")
+				.required(true)
+				.repeatable(false)
+				.argument("dir")
+				.callback(OptionCallback<DBLogger>(this, &DBLogger::handleLogDirectory)));
+
+		options.addOption(
+			Option("connector", "c", "database connector")
+				.required(true)
+				.repeatable(false)
+				.argument("connector")
+				.callback(OptionCallback<DBLogger>(this, &DBLogger::handleDbConnector)));
+
+		options.addOption(
+			Option("constring", "s", "database connection string")
+				.required(true)
+				.repeatable(false)
+				.argument("cstr")
+				.callback(OptionCallback<DBLogger>(this, &DBLogger::handleDbConnectionString)));
+
+	}
+
+	void handleHelp(const std::string& name, const std::string& value)
+	{
+		_helpRequested = true;
+		displayHelp();
+		stopOptionsProcessing();
+	}
+
+	void handleLogDirectory(const std::string& name, const std::string& value)
+	{
+		if (value.empty())
+			throw Poco::Util::IncompatibleOptionsException("Expected directory name");
+		_directory = value;
+	}
+
+	void handleDbConnector(const std::string& name, const std::string& value)
+	{
+		if (value.empty())
+			throw Poco::Util::IncompatibleOptionsException("Expected database connector name");
+		_connector = value;
+	}
+
+	void handleDbConnectionString(const std::string& name, const std::string& value)
+	{
+		if (value.empty())
+			throw Poco::Util::IncompatibleOptionsException("Expected database connection string");
+		_connectionString = value;
+	}
+
+	void displayHelp()
+	{
+		HelpFormatter helpFormatter(options());
+		helpFormatter.setCommand(commandName());
+		helpFormatter.setUsage("OPTIONS");
+		helpFormatter.format(std::cout);
+	}
+
+	void onItemAdded(const Poco::DirectoryWatcher::DirectoryEvent& ev)
+	{
+		logger().trace("Added: %s", ev.item.path());
+		processFile(ev.item);
+	}
+
+	void onItemModified(const Poco::DirectoryWatcher::DirectoryEvent& ev)
+	{
+		logger().information("Modified: %s", ev.item.path());
+		processFile(ev.item);
+	}
+
+	void onItemMovedFrom(const Poco::DirectoryWatcher::DirectoryEvent& ev)
+	{
+		logger().trace("Moved from: %s", ev.item.path());
+		processFile(ev.item);
+	}
+
+	void onItemMovedTo(const Poco::DirectoryWatcher::DirectoryEvent& ev)
+	{
+		logger().trace("Moved to: %s", ev.item.path());
+	}
+
+	void onItemRemoved(const Poco::DirectoryWatcher::DirectoryEvent& ev)
+	{
+		logger().trace("Removed: %s", ev.item.path());
+	}
+
+	void processFile(const Poco::File& file)
```

Comment:
The entire processing sequence should be broken down into (a) detection, (b) forwarding to DB and (c) deleting the file or marking it as error. This should be done in threads (or `Poco::Task`s), with one scanner doing (a) and multiple (number configurable) workers doing (b),(c). 

So, let's leave just the Application functionality  in the main thread, then have (1) one (optional) sample message creation thread, (2) one thread just for monitoring the directory and (3) multiple worker threads for processing files. 

The monitoring thread just detects new files, and puts the file names into a shared `NotificationQueue`. Working threads are removing paths from the queue and doing the actual work. Number of working threads should be configurable so app can scale. Working threads should be capable to graciously deal with the condition when file does not exist (eg. file deleted while notification was waiting in the queue).

Reply:
With exception of initial directory scan the changes in the directory are monitored with DirectoryWatcher. So far it seems that it works properly.

Will add configurable workers to populate the database.

Reply:
Reimplemented to use DirectoryIterator and NotificationQueue, however something does not work right because processing gets stuck after few iterations.

I'll find a different way to process the files.

Reply:
Another refactoring using STL.